sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -125,22 +125,20 @@ class Analyzer(
   object CTESubstitution extends Rule[LogicalPlan] {
     // TODO allow subquery to define CTE
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
-      case With(child, relations) => substituteCTE(child, relations)
+      case With(child, relations) if relations.nonEmpty =>
+        substituteCTE(child, relations.foldLeft(Seq.empty[(String, LogicalPlan)])((resolved, r) =>
+          resolved ++ Seq(r._1 -> ResolveRelations(substituteCTE(r._2, resolved)))))
```

Comment:
`resolved :+ r._1 -> ResolveRelations(substituteCTE(r._2, resolved))`?

You could also deconstruct the `r` tuple for easier reading.